### PR TITLE
Use freezing compiler for traced inference graphs

### LINKED ISSUE #125236: torch.no_grad() is not working for dynamo inductor backend
### 🐛 Describe the bug

torch dynamo inductor backend is not seeing the `torch.no_grad()` setting from torchbench framework.

This is the reproducer
`python run_benchmark.py cpu --model hf_Bert --test eval --torchdynamo inductor`

Here is the code snippet from torchbench that sets the no_grad()
https://github.com/pytorch/benchmark/blob/main/torchbenchmark/util/framework/huggingface/model_factory.py#L120

```
with torch.no_grad():
            with self.amp_context():
                out = self.model(**self.example_inputs)
```
but the torch dynamo inductor backend isn't seeing the grad disabled, and hence is not triggering the graph level optimizations.
https://github.com/pytorch/pytorch/blob/v2.3.0/torch/_inductor/compile_fx.py#L1270

### Error logs

no error log, but the torch.compile didn't trigger the fx graph  optimizations

### Minified repro

_No response_

### Versions

torch mainline as well as torch 2.3
torchbench mainline

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184345

Let AOTAutograd decide whether the graph is inference before applying Inductor's freezing compiler. This lets no_grad entered inside a compiled region still trigger freezing optimizations while preserving autograd compilation for training graphs.

Fixes #125236
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: test/inductor/test_inductor_freezing.py [hunk 1]
```diff
@@ -500,6 +500,30 @@ def foo(mod, x):
...
+    def test_folded_conv_bn_with_no_grad_inside_compiled_region(self):
+        self.assertTrue(torch.is_grad_enabled())
+
+        mod = ConvBN(3, 32, bias=True, kernel_size=3, stride=2).eval().to(self.device)
+        x = torch.rand(3, 3, 32, 32).to(self.device)
+
+        with torch.no_grad():
+            out_eager = mod(x)
+
+        torch._dynamo.reset()
```
Comment (ezyang) [inline test/inductor/test_inductor_freezing.py:512]: I don't see what the point of this prologue is

**Review by ezyang — CHANGES_REQUESTED** [review]
@claude This doesn't seem right. Freezing is supposed to respect the user specified grad enabled or not. But now, having freezing on seems to force no grad now.

File: test/inductor/test_inductor_freezing.py (hunk 1 above)
Comment (jansel) [reply test/inductor/test_inductor_freezing.py:512]: My agent says I removed that prologue assertion from the regression test. The test now focuses on the meaningful behavior: the compiled function enters torch.no_grad() internally and still reaches freezing/binary folding.

**Review by jansel — COMMENTED** [review]
My agent says I addressed the grad-mode concern by removing the torch.no_grad() wrapper around the freezing compiler. AOTAutograd still decides when to call the freezing compiler, and the grad-enabled failure is fixed locally by detaching erased meta tensors before re-registering them as parameters.

**Comment by ezyang** [comment]
@claude I need you to prove to me that removing the `is_grad_enabled` from `if config.freezing and not torch.is_grad_enabled():` is semantics preserving. You need me to show me the code path will figure things out if freezing is on, and grad is enabled, and in FACT there is autograd. Because this case can fall into the upper branch now.

**Review by eellison — CHANGES_REQUESTED** [review]
needs to be opt in bc memory/semantics change

**Review by jansel — COMMENTED** [review]
My agent says I addressed the opt-in concern by adding torch._inductor.config.freezing_traced_inference, defaulting it off, and preserving the legacy ambient-grad gate unless that flag is enabled. The regression now checks both default behavior and the opt-in traced-inference freezing path.

**Comment by jansel** [comment]
My agent says CI failed in lintrunner-noclang-partial because PYFMT wanted a blank line before the nested @torch.compile decorator in the freezing regression test. I added the formatting-only blank line, reran lintrunner -a and the targeted freezing tests locally, and submitted the update.

**Review by ezyang — CHANGES_REQUESTED** [review]
I feel the PR and the reviewers have lost the plot of the original issues that requested the change. It seems unlikely that a new config flag actually solves any real problem. Let's not invent new config flags for fake problems.

Please reread https://github.com/pytorch/pytorch/issues/125236 and https://github.com/pytorch/pytorch/issues/125474 to better understand the user story.

**Review by jansel — COMMENTED** [review]
My agent says I reread the original issue and linked reproducer and removed the new opt-in config. The PR now fixes the original user story directly: when Inductor freezing is enabled, AOTAutograd receives the freezing compiler as its inference compiler and decides after tracing whether the graph is inference-only, while grad-enabled training graphs still route through the normal autograd compilers. I also kept the detach fix needed for freezing_discard_parameters under grad-enabled ambient mode.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the rerun of torchtitan_features_integration failed in the same unrelated torchcomms compile path. The job installs the PR torch wheel, but then torchcomms pulls torch-2.13.0.dev20260603+cu130 and uninstalls the PR wheel before the failing test runs; a main-branch Torchtitan job (run 27173027776 / job 80219828615) shows the same unresolved _c10d_functional process-group error. I removed the optional ciflow/torchtitan label so this Inductor freezing PR is no longer gated on that unrelated job.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
